ingestion-beam/src/main/java/com/mozilla/telemetry/io/Read.java

```diff
@@ -64,11 +68,33 @@ public FileInput(ValueProvider<String> fileSpec, InputFileFormat fileFormat) {
 
     @Override
     public Result<PCollection<PubsubMessage>> expand(PBegin input) {
-      return input.apply(TextIO.read().from(fileSpec)).apply(fileFormat.decode());
+      return input.apply(TextIO.read().from(fileSpec)).apply(fileFormat.decode()).apply(Flatten);
     }
   }
 
-  /** Implementation of reading from local or remote files. */
+  /** Implementation of reading from heka blobs stored as files */
+  public static class HekaInput extends Read {
+
+    private final ValueProvider<String> fileSpec;
+
+    public HekaInput(ValueProvider<String> fileSpec) {
+      this.fileSpec = fileSpec;
+    }
+
+    @Override
+    public Result<PCollection<PubsubMessage>> expand(PBegin input) {
+      return input.apply(FileIO.match().filepattern(fileSpec)).apply(FileIO.readMatches()).apply(new HekaTransform())
```

Comment:
`.apply(new HekaTransform())` wraps its output in Result, so it looks like the type here would be incorrect for the flatmap. We need to pass just the successful output into the flatmap. I'm playing with a solution now.

Reply:
```
      Result<PCollection<List<PubsubMessage>>> result = input
          .apply(FileIO.match().filepattern(fileSpec)).apply(FileIO.readMatches())
          .apply(new HekaTransform());
      return Result.of(result.output().apply(Flatten.iterables()), result.errors());
```

Reply:
Makes perfect sense! It's difficult to keep track of all these types 😢

Reply:
Looks like Beam can't automatically infer the type here, so this could become:

```
      Result<PCollection<List<PubsubMessage>>> result = input
          .apply(FileIO.match().filepattern(fileSpec)).apply(FileIO.readMatches())
          .apply(new HekaTransform());
      result.output().setCoder(ListCoder.of(PubsubMessageWithAttributesCoder.of()));
      return Result.of(result.output().apply(Flatten.iterables()), result.errors());
```

where we have an explicit `setCoder` call.

Reply:
Thanks! That did it